### PR TITLE
KUBOS-131 HAL Unit Tests + misc

### PR DESCRIPTION
**Pull request by CatFreed**
Created unit tests for:
- UART
- I2C
- SPI

Other changes:
- fixed i2c/spi bus off-by-one error
- added configuration validation to reject bad config settings
- added support for more baud rates
- added support for write_immediate and write_immediate_str
- updated doxygen docs
- added support for KUARTStatus

### CODE REVIEW

File: source/uart.c [hunk 1]
```diff
@@ -213,6 +241,43 @@ void hal_uart_interrupt(hal_uart_handle * handle)
...
+/**
+ * Write a character directly to the uart interface
+ * @param uart uart bus
+ * @param c character to write
+ * @return KUARTStatus UART_OK if success, otherwise failure
+ */
+KUARTStatus k_uart_write_immediate(KUARTNum uart, char c)
+{
+    hal_uart_handle *handle = uart_handle(uart);
+    if (!handle) {
```
Comment (rplauche) [inline source/uart.c]: nit: This null check uses `(!handle)`, checks earlier in the file use `(handle == NULL)`

File: test/i2c.c [hunk 2]
```diff
@@ -0,0 +1,641 @@
+/*
+ * KubOS Core Flight Services
+ * Copyright (C) 2015 Kubos Corporation
```
Comment (rplauche) [inline test/i2c.c]: nit: Copyright year should be 2016

File: test/i2c.c [hunk 3]
```diff
@@ -0,0 +1,641 @@
...
+
+    ret = kprv_i2c_dev_init(K_NUM_I2CS+1);
+    TEST_ASSERT_EQUAL_INT_MESSAGE(I2C_ERROR_NULL_HANDLE, ret, "Successfully initialized fake i2c port?");
+
+}
+
+/**
+ * test_i2c_termInit
+ *
+ * Purpose:  Test terminating a port that wasn't initialized
```
Comment (rplauche) [inline test/i2c.c]: It looks like the purpose here is to test terminating a port which has been initialized?

File: test/i2c.c [hunk 4]
```diff
@@ -0,0 +1,641 @@
...
+        .clock_speed = 100000
+    };
+
+    KI2C *k_i2c = kprv_i2c_get(I2C_BUS);
+    memcpy(&k_i2c->conf, &conf, sizeof(KI2CConf));
+    k_i2c->bus_num = I2C_BUS;
+    k_i2c->i2c_lock = xSemaphoreCreateMutex();
+
+    ret = kprv_i2c_dev_init(I2C_BUS);
+    TEST_ASSERT_EQUAL_INT_MESSAGE(I2C_ERROR, ret, "Successfully inititalized I2C_BUS?");
```
Comment (rplauche) [inline test/i2c.c]: Should this be `I2C_ERROR` or `I2C_ERROR_CONFIG`?

File: test/spi.c [hunk 5]
```diff
@@ -0,0 +1,913 @@
+/*
+ * KubOS Core Flight Services
+ * Copyright (C) 2015 Kubos Corporation
```
Comment (rplauche) [inline test/spi.c]: nit: Copyright year should be 2015

File: test/spi.c [hunk 6]
```diff
@@ -0,0 +1,913 @@
...
+static void test_spi_termNoninit(void)
+{
+    int ret;
+
+    ret = kprv_spi_dev_terminate(SPI_BUS);
+    TEST_ASSERT_EQUAL_INT_MESSAGE(SPI_OK, ret, "Failed to terminate SPI_BUS");
+}
+
+/*
+ * test_spi_termNoninit
```
Comment (rplauche) [inline test/spi.c]: nit: Wrong test name

File: test/spi.c [hunk 7]
```diff
@@ -0,0 +1,913 @@
...
+{
+    int ret;
+    uint8_t chipReg = 0xD0; //Chip ID register
+    chipReg |= 0x80; //Turn on high bit for read request
+
+    test_spi_setup();
+
+    k_gpio_write(CS, 0);
+
+    ret = kprv_spi_write(SPI_BUS, &chipReg, 100);
```
Comment (rplauche) [inline test/spi.c]: Haha how does this not wreck the system? We don't appear to have any overflow protection in the write functions (insert future custom buffer system issue here)

File: test/uart.c [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+
+    TEST_ASSERT_FALSE(k_uart_init(uart_bus, &conf));
+
+    vTaskDelay(50);
+
+    returnLenRead = k_uart_read(uart_bus, buffer, 100);
+
+    k_uart_terminate(uart_bus);
+
+    TEST_ASSERT_EQUAL_INT_MESSAGE(len, returnLenWrite, "Failed to write");
```
Comment (rplauche) [inline test/uart.c]: Why do you test the write len if nothing is written

File: source/i2c.c [hunk 9]
```diff
@@ -98,13 +99,23 @@ KI2CStatus kprv_i2c_dev_init(KI2CNum i2c)
...
     }
 
+    /*
+     * We don't currently support slave mode.  Initialization should fail
+     * if it's specified.
+     * Remove this once slave mode is implemented.
+     */
+    if(k_i2c->conf.role == K_SLAVE)
+    {
+        return I2C_ERROR;
```
Comment (rplauche) [inline source/i2c.c]: Should this be `I2C_ERROR_CONFIG`?

File: test/spi.c (hunk 6 above)
Comment (CatFreed) [inline test/spi.c]: No it's not?

**Pull request merged**